src/shaka_scheme/system/base/PrimitiveFormMarker.cpp

```diff
@@ -0,0 +1,53 @@
+//
+// Created by aytas on 11/8/2017.
+//
+
+#include "shaka_scheme/system/base/PrimitiveFormMarker.hpp"
+
+namespace shaka {
+
+PrimitiveFormMarker::PrimitiveFormMarker(std::string str) :
+    str(str) {}
+
+PrimitiveFormMarker::PrimitiveFormMarker(const PrimitiveFormMarker& other) :
+    str(other.str) {}
+
+PrimitiveFormMarker::PrimitiveFormMarker(PrimitiveFormMarker&& other) :
+    str(std::move(other.str)) {}
+
+PrimitiveFormMarker& PrimitiveFormMarker::operator=(PrimitiveFormMarker other) {
+  PrimitiveFormMarker temp(other);

```

Comment:
Since you are taking in the other PrimitiveFormMarker by value, isn't it already implicitly copied? Not sure if this line is needed but I suppose it doesn't hurt necessarily.

Reply:
This is part of the [copy-and-swap idiom](https://stackoverflow.com/questions/3279543/what-is-the-copy-and-swap-idiom) that is commonly used when you need to implement the copy/move constructors and assignment operators -- this is called the [Rule of Five](https://stackoverflow.com/questions/4782757/rule-of-three-becomes-rule-of-five-with-c11) or Rule of Three with pre-C++11.

Reply:
Right so, in the case of an rvalue having been passed in, the `other` parameter will be constructed via the move constructor which you have defined, so we are free to swap from `other` without fear. In the case of an lvalue having been passed in, the `other` parameter will be constructed via the copy constructor which you have also defined, so we again are free to go ahead and swap directly from `other`. No need for the creation of `temp` in either case if I'm not mistaken. But no biggie.

Reply:
I've relooked over things -- you're right. The change should be made -- maybe a new pull request should be made.